spec/requests/session_controller_spec.rb

```diff
@@ -2030,6 +2030,71 @@ def login_with_sso_and_invite(invite_key = invite.invite_key)
         end
       end
 
+      describe "when user's password has been marked as expired" do
+        let!(:expired_user_password) do
+          Fabricate(
+            :expired_user_password,
+            user:,
+            password: "myawesomepassword",
+            password_salt: user.salt,
+            password_algorithm: user.password_algorithm,
+          )
+        end
+
+        before { RateLimiter.enable }
+
+        use_redis_snapshotting
+
+        it "should return an error response code with the right error message and enqueues the password reset email" do
+          expect_enqueued_with(
+            job: :critical_user_email,
+            args: {
+              type: "forgot_password",
+              user_id: user.id,
+            },
+          ) do
+            post "/session.json", params: { login: user.username, password: "myawesomepassword" }
+          end
+
+          expect(response.status).to eq(200)
+          expect(session[:current_user_id]).to eq(nil)
+        end
+
+        it "should limit the number of forgot password emails sent a day to the user when logging in with an expired password" do
+          Discourse.redis.flushdb
```

Comment:
I have to flush here because there is some state being leaked that is affecting the rate limits triggered by this test. I don't know what is causing the leak at the moment.

Reply:
this is intense though, why not just junk the 1/2 keys you need up front? expose a static method in controller to do so...

Reply:
I can do that too but was rushing a little here. Will have a closer look.

Reply:
Interesting, I no longer have this problem locally.. maybe my local Redis DB state was bad just now.